### ISSUE TITLE
Realm: OOM error when running on recent versions of the runtime

### ISSUE THREAD

**Issue by mariodirenzo** [StanfordLegion/legion#1513]
I'm experiencing deterministic OOM errors when running HTR on recent versions of the runtime. By doing some bisections I've seen that merge of caching allocator could be the first commit when I reproduce the issue.
If I revert the merge with `git revert -m 1 6d2e2bc1` on the latest control_replication (`1b22ee27122bdea14cedf9373b1433d4741aa381`), the problem goes away.

**Comment by streichler** [comment]
@mariodirenzo can you see if changing this define (without reverting the entire commit) eliminates the problem as well?
```
diff --git a/runtime/realm/realm_config.h b/runtime/realm/realm_config.h
index d79fafe88..d0b9b1032 100644
--- a/runtime/realm/realm_config.h
+++ b/runtime/realm/realm_config.h
@@ -72,7 +72,7 @@
 
 // ASAN has issues with thread local destructors, so disable this path for asan
 #if !defined(REALM_USE_CACHING_ALLOCATOR) && !defined(ASAN_ENABLED)
-  #define REALM_USE_CACHING_ALLOCATOR 1
+  #define REALM_USE_CACHING_ALLOCATOR 0
   #ifndef REALM_TASK_BLOCK_SIZE
     #define REALM_TASK_BLOCK_SIZE (128ULL*1024ULL)
   #endif
```

There's another issue being debugged that looks to be related to this commit, and the above patch helped there.

**Comment by mariodirenzo** [comment]
Yes, it does

**Comment by muraj** [comment]
@mariodirenzo Thanks for reporting this, I'm submitting a PR to temporarily disable the caching allocator in the meantime while I debug this.  Can I ask, how many unique CPU threads do you have calling into realm to spawn tasks at any point in the app?  The new caching allocator will allocate a fairly large amount of memory for each unique thread that is spawning tasks, but it should reuse these blocks and not create more.  I'll write up a local test app to verify these paths real quick and see if I can't debug this better.

**Comment by elliottslaughter** [comment]
> Can I ask, how many unique CPU threads do you have calling into realm to spawn tasks at any point in the app?

I'm pretty sure this is a function of the degree of concurrency in Legion runtime meta tasks, and not something most users would know (or be aware of). It's not even really a function of the `-ll:util` count because Realm creates potentially many threads for a single processor.

What you'd want (I assume) would be to count the threads associated with Legion utility processors, but @lightsighter will need to confirm if that's sufficient or if we need to include other processors as well.

**Comment by lightsighter** [comment]
> It's not even really a function of the -ll:util count because Realm creates potentially many threads for a single processor.

Yes, but even that is bounded by a small integer (unless you're running with `-ll:force_kthreads`). Even if there are lots of tasks blocked waiting on events, those will just consume lots of stack memory because of Realm's user-level threading. There's still a fixed number of Realm threads running which will bound the number thread-local free lists.

**Comment by mariodirenzo** [comment]
@muraj thanks for creating the PR! As anticipated by Elliott and Mike, I don't have any idea how many CPU threads are called by the application. However, I am happy to execute the code with specific logs/flags that can provide this information.

**Comment by mariodirenzo** [comment]
I can provide a reproducing sample on sapling2. Do you have access to the system?
In case, I can provide a simple set of instructions.
I've also noticed that the flag `-ll:force_kthreads` is required in order to reproduce the issue. I do not know if that tells you something about the problem.

**Comment by lightsighter** [comment]
If @muraj does not already have an account on `sapling2` then I will make him one. Go ahead and make a reproducer there.

**Comment by muraj** [comment]
@mariodirenzo I uploaded a PR in the cperry/caching-allocator-fix branch to re-enable the caching allocator, but keep it disabled when force_kthreads is enabled, and also greatly reduced the block sizes (from 128k to 4k).  You're welcome to try it if you'd like, or if you want to make the modifications yourself to just reduce the block sizes and see how the memory usage changes?  It shouldn't leak if the kthread workers are bounded, but if they're unbounded, then we're going to have a lot of threads, each owning 128k * 500B of memory, which will probably eat up your RAM pretty quickly.

**Comment by lightsighter** [comment]
@muraj One thing that @streichler and I discussed a while ago was having some kind of warning being reported whenever a large number of preempted tasks were being created which necessitated many stacks with user-level threading being created or many kernel threads  being created when running with kthreads. Usually whenever this happens it's indicative of a performance issue in the client program and it would probably be good to investigate it. The threshold for the warning can be configurable so clients can set it as desired. It would have the added benefit here of suggesting what is going wrong even before we run out of memory. Does that seem like something we can add?

**Comment by mariodirenzo** [comment]
@muraj, thanks for uploading the PR. I've locally merged it with control_replication and confirmed that it fixes the issue with and without `force_kthreads`. It would be hard for me to set the memory usage properly but I agree with Mike that a warning issued by the runtime could help identify the problem.

**Comment by muraj** [comment]
I believe this issue is resolve as stated, so moving this to closed.  If additional debugging information is required outside what is already available today, please file another issue tracking this request.  If the issue is not resolved as stated, please reopen.
